include/Gaffer/Context.h

```diff
@@ -275,14 +275,16 @@ class GAFFER_API Context : public IECore::RefCounted
 			// And use this ownership flag to tell us when we need to do explicit
 			// reference count management.
 			Ownership ownership;
+			// This tells us the hash value we need to subtract off the incremental hash if we change this value
+			IECore::MurmurHash hash;
 		};
 
+		void validateHashes();
+
 		typedef boost::container::flat_map<IECore::InternedString, Storage> Map;
 
 		Map m_map;
 		ChangedSignal *m_changedSignal;
-		mutable IECore::MurmurHash m_hash;
-		mutable bool m_hashValid;
```

Comment:
Yeah, you're totally right - these were ripped out for V1, where they really weren't necessary any more, but for the approach we've decided on, they probably are needed again.

Reply:
One weird corner case I hesitated on ... in the previous code, calling `changed( "variableNameThatDoesntExist" )` would set m_hashValid to false.  Now that I'm actually finding the entry for the hash value ( so that I can recompute the per-entry hash ), it feels more natural and correct to only invalidate m_hashValid if the name refers to a variable that actually exists in the context?  Any opinions on this change in behaviour?

Reply:
Fixed in 7e95c0f

Reply:
> it feels more natural and correct to only invalidate m_hashValid if the name refers to a variable that actually exists in the context? Any opinions on this change in behaviour?

Yep, sounds good.